src/proc/constants.rs

```diff
@@ -0,0 +1,564 @@
+                }
+            }
+            Expression::ArrayLength(expr) => {
+                let array = self.solve(expr)?;
+
+                match self.constants[array].inner {
+                    crate::ConstantInner::Scalar { .. } => {
+                        Err(ConstantSolvingError::InvalidArrayLengthArg)
+                    }
+                    crate::ConstantInner::Composite { ty, .. } => match self.types[ty].inner {
+                        crate::TypeInner::Array { size, .. } => match size {
+                            crate::ArraySize::Constant(constant) => Ok(constant),
+                            crate::ArraySize::Dynamic => {
+                                Err(ConstantSolvingError::ArrayLengthDynamic)
+                            }
+                        },
+                        _ => Err(ConstantSolvingError::InvalidArrayLengthArg),
+                    },
+                }
+            }
+
+            Expression::Load { .. } => Err(ConstantSolvingError::Load),
+            Expression::Select { .. } => Err(ConstantSolvingError::Select),
+            Expression::LocalVariable(_) => Err(ConstantSolvingError::LocalVariable),
+            Expression::Derivative { .. } => Err(ConstantSolvingError::Derivative),
+            Expression::Relational { .. } => Err(ConstantSolvingError::Relational),
+            Expression::Call { .. } => Err(ConstantSolvingError::Call),
+            Expression::FunctionArgument(_) => Err(ConstantSolvingError::FunctionArg),
+            Expression::GlobalVariable(_) => Err(ConstantSolvingError::GlobalVariable),
+            Expression::ImageSample { .. } => Err(ConstantSolvingError::ImageExpression),
+            Expression::ImageLoad { .. } => Err(ConstantSolvingError::ImageExpression),
+        }
+    }
+
+    fn access(
+        &mut self,
+        base: Handle<Expression>,
+        index: usize,
+    ) -> Result<Handle<Constant>, ConstantSolvingError> {
+        let base = self.solve(base)?;
+
+        match self.constants[base].inner {
+            crate::ConstantInner::Scalar { .. } => Err(ConstantSolvingError::InvalidAccessBase),
+            crate::ConstantInner::Composite { ty, ref components } => match self.types[ty].inner {
+                crate::TypeInner::Vector { size, .. } => {
+                    if size as usize <= index {
+                        Err(ConstantSolvingError::InvalidAccessIndex)
+                    } else {
+                        Ok(components[index])
+                    }
+                }
+                crate::TypeInner::Matrix {
+                    rows,
+                    columns,
+                    width,
+                } => {
+                    if columns as usize <= index {
+                        return Err(ConstantSolvingError::InvalidAccessIndex);
+                    }
+
+                    let components =
+                        components[(index * rows as usize)..((index + 1) * rows as usize)].to_vec();
```

Comment:
This might be a problem. The composite components for a matrix in our IR should be vectors and only vectors.
But this code works with components as if they were scalars. I assume that's because when implementing matrix construction in GLSL it was easier to just fill out the components vector right away? The IR will fail validation on that once the validation matures a bit.

Reply:
I did this way because I thought they were supposed to be flattened should have asked